sessionctx/variable/setvar_test.go

```diff
@@ -0,0 +1,155 @@
+		{"tidb_merge_join_concurrency", []string{"1", "2"}},
+		{"tidb_index_merge_intersection_concurrency", []string{"1", "2"}},
+		{"tidb_opt_projection_push_down", []string{"0", "1"}},
+		{"tidb_enable_vectorized_expression", []string{"0", "1"}},
+		{"tidb_opt_join_reorder_threshold", []string{"0", "1"}},
+		{"tidb_enable_index_merge", []string{"0", "1"}},
+		{"tidb_enable_extended_stats", []string{"0", "1"}},
+		{"tidb_executor_concurrency", []string{"1", "2"}},
+		{"tidb_enable_index_merge_join", []string{"0", "1"}},
+		{"tidb_enable_ordered_result_mode", []string{"0", "1"}},
+		{"tidb_enable_pseudo_for_outdated_stats", []string{"0", "1"}},
+		{"tidb_stats_load_sync_wait", []string{"0", "1"}},
+		{"tidb_cost_model_version", []string{"1", "2"}},
+		{"tidb_index_join_double_read_penalty_cost_rate", []string{"0", "1"}},
+		{"tidb_default_string_match_selectivity", []string{"0", "1"}},
+		{"tidb_enable_prepared_plan_cache", []string{"0", "1"}},
+		{"tidb_enable_non_prepared_plan_cache", []string{"0", "1"}},
+		{"tidb_plan_cache_max_plan_size", []string{"0", "1"}},
+		{"tidb_opt_range_max_size", []string{"0", "1"}},
+		{"tidb_opt_advanced_join_hint", []string{"0", "1"}},
+		{"tidb_opt_prefix_index_single_scan", []string{"0", "1"}},
+		{"tidb_store_batch_size", []string{"1", "2"}},
+		{"tidb_enable_inl_join_inner_multi_pattern", []string{"0", "1"}},
+		{"tidb_opt_enable_late_materialization", []string{"0", "1"}},
+		{"tidb_opt_ordering_index_selectivity_threshold", []string{"0", "1"}},
+		{"tidb_opt_enable_mpp_shared_cte_execution", []string{"0", "1"}},
+	}
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	for _, c := range testCases {
+		for _, setv := range c.cases {
+			tk.MustQuery(fmt.Sprintf("select /*+ set_var(%v=%v) */ @@%v", c.varName, setv, c.varName)).Check(testkit.Rows(setv))
+			r := tk.MustQuery(fmt.Sprintf("select @@%v", c.varName)).Rows()
+			tk.MustExec(fmt.Sprintf("set @@%v=default", c.varName))
+			tk.MustQuery(fmt.Sprintf("select @@%v", c.varName)).Check(r)
+		}
+	}
+}
+
+func TestSetVarStringOrEnum(t *testing.T) {
+	testCases := []struct {
+		varName string
+		cases   []string
+	}{
+		{"tidb_read_consistency", []string{"strict", "weak"}},
+		{"tidb_isolation_read_engines", []string{"tidb,tiflash", "tikv,tidb"}},
+		{"tidb_replica_read", []string{"follower", "prefer-leader"}},
+		{"tidb_partition_prune_mode", []string{"static", "dynamic"}},
+		{"mpp_version", []string{"0", "1"}},
+		{"tidb_opt_fix_control", []string{"44262:ON", "44389:ON,44823:ON"}},
+		// {"tidb_runtime_filter_type", []string{}},
+		{"tidb_runtime_filter_mode", []string{"OFF", "LOCAL"}},
+		{"sql_mode", []string{"", "ONLY_FULL_GROUP_BY"}},
+	}
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	for _, c := range testCases {
+		for _, setv := range c.cases {
+			tk.MustQuery(fmt.Sprintf("select /*+ set_var(%v=\"%v\") */ @@%v", c.varName, setv, c.varName)).Check(testkit.Rows(setv))
```

Comment:
It seems that this UT only test the ```set_var``` does not affect the variable. But it does not test whether the variable takes effect in the query by ```set_var```.
So wouldn't it be better to add such checks?

Reply:
There's no better idea to test each of them at one time. I think we can only pick some of them to test in the way you said.

Reply:
OK, just a suggestion. (After all, if all the changeable variables are tested this time, the amount of work will be a bit large.)